lib/resty/hmac.lua

```diff
@@ -0,0 +1,94 @@
+-- Adds HMAC support to Lua with multiple algorithms, via OpenSSL and FFI
+--
+-- Author: ddragosd@gmail.com
+-- Date: 16/05/14
+--
+
+
+local ffi = require "ffi"
+local ffi_new = ffi.new
+local ffi_str = ffi.string
+local C = ffi.C
+local resty_string = require "resty.string"
+local setmetatable = setmetatable
+local error = error
+
+
+local _M = { _VERSION = '0.09' }
+
+
+local mt = { __index = _M }
+
+--
+-- EVP_MD is defined in openssl/evp.h
+-- HMAC is defined in openssl/hmac.h
+--
+ffi.cdef[[
+typedef struct env_md_st EVP_MD;
+typedef struct env_md_ctx_st EVP_MD_CTX;
+unsigned char *HMAC(const EVP_MD *evp_md, const void *key, int key_len,
+		    const unsigned char *d, size_t n, unsigned char *md,
+		    unsigned int *md_len);
+const EVP_MD *EVP_sha1(void);
+const EVP_MD *EVP_sha224(void);
+const EVP_MD *EVP_sha256(void);
+const EVP_MD *EVP_sha384(void);
+const EVP_MD *EVP_sha512(void);
+]]
+
+-- table definind the available algorithms and the length of each digest
+-- for more information @see: http://csrc.nist.gov/publications/fips/fips180-4/fips-180-4.pdf
+local available_algorithms = {
+    sha1   = { alg = C.EVP_sha1(),   length = 160/8  },
+    sha224 = { alg = C.EVP_sha224(), length = 224/8  },
+    sha256 = { alg = C.EVP_sha256(), length = 256/8  },
+    sha384 = { alg = C.EVP_sha384(), length = 384/8  },
+    sha512 = { alg = C.EVP_sha512(), length = 512/8  }
+}
+
+-- 64 is the max lenght and it covers up to sha512 algorithm
+local digest_len = ffi_new("int[?]", 64)
+local buf = ffi_new("char[?]", 64)
```

Comment:
Is this declaring a single buffer instance that gets reused across all instances of hmac? I'm new to Lua, but if this is the case, that would be dangerous (wouldn't it?).

Reply:
@ljwagerfield This is a temporary buffer that is used inside a single Lua function that never yields. Since nginx uses single threaded processes and it does not yield for Lua light threads, it is safe.